lib/internal/modules/esm/default_resolve.js

```diff
@@ -75,15 +76,15 @@ function resolve(specifier, parentURL) {
       }
```

Comment:
I think the simpler fix would just be a single line here on line 77:

```suggestion
      }
      if (!format) {
        format = asyncESM.typeFlag;
      }
```

Note also that the `type === 0` check you have done is effectively saying we shouldn't be able to override when there is an existing "type" in the package.json. If that is what we want, we could explicitly change the `if (ext === '.js')` check in the validations above to `if (ext !== '.cjs' && ext !== '.mjs')` so that we get a proper validation error in this case.